src/samplers/cpu/perf.c

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019-2020 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+#include <linux/ptrace.h>
+#include <uapi/linux/bpf_perf_event.h>
+
+// Arrays which hold the perf counters
+BPF_PERF_ARRAY(branch_instructions_array, NUM_CPU);
+BPF_PERF_ARRAY(branch_misses_array, NUM_CPU);
+BPF_PERF_ARRAY(cache_misses_array, NUM_CPU);
+BPF_PERF_ARRAY(cache_references_array, NUM_CPU);
+BPF_PERF_ARRAY(cycles_array, NUM_CPU);
+BPF_PERF_ARRAY(dtlb_load_miss_array, NUM_CPU);
+BPF_PERF_ARRAY(dtlb_load_access_array, NUM_CPU);
+BPF_PERF_ARRAY(dtlb_store_miss_array, NUM_CPU);
+BPF_PERF_ARRAY(dtlb_store_access_array, NUM_CPU);
+BPF_PERF_ARRAY(instructions_array, NUM_CPU);
+BPF_PERF_ARRAY(reference_cycles_array, NUM_CPU);
+
+// Tables which are read in user space
+BPF_ARRAY(branch_instructions, u64, NUM_CPU);
+BPF_ARRAY(branch_misses, u64, NUM_CPU);
+BPF_ARRAY(cache_misses, u64, NUM_CPU);
+BPF_ARRAY(cache_references, u64, NUM_CPU);
+BPF_ARRAY(cycles, u64, NUM_CPU);
+BPF_ARRAY(dtlb_load_access, u64, NUM_CPU);
+BPF_ARRAY(dtlb_load_miss, u64, NUM_CPU);
+BPF_ARRAY(dtlb_store_access, u64, NUM_CPU);
+BPF_ARRAY(dtlb_store_miss, u64, NUM_CPU);
+BPF_ARRAY(instructions, u64, NUM_CPU);
+BPF_ARRAY(reference_cycles, u64, NUM_CPU);
+
+int do_count(struct bpf_perf_event_data *ctx) {
+    u32 cpu = bpf_get_smp_processor_id();
+    u64 count = 0;
+
+    count = branch_instructions_array.perf_read(CUR_CPU_IDENTIFIER);
```

Comment:
> map.perf_read() will return a hardware-calculated counter of the number of cycles elapsed from some point in the past 

Not sure whether "some point in the past" counters are _good enough_ measurements. Although I like doing less work inside the bpf handler.

Reply:
Interesting. I think this would still be okay - most counters that Rezolus reads have some initial value starting from an arbitrary epoch. Typically, we would either be doing a rate of the counter value taken from the metrics endpoint, or relying on the histogram readings which already calculate a normalized rate. I don't think the aim is to capture the exact number of 'instructions' retired (or whatever metric) since Rezolus launched, but to provide insight into the rates at which these events happen.

Given that, do you think this approach is acceptable? I do like having the bpf handler be more lightweight.

Reply:
Yeah I totally agree. 
My only concern was some misbehaving counter not updating for a some time thus skewing the total counter. But, I dont think this skew would be significant enough to warrant a heavier bpf program.

👍

Reply:
This gave me another thought about the approach here that I need to think about a bit more. Didn't see it happen in testing yet, but since the frequency of the underlying perf sampling happens at the same frequency as the sampler executing - any frequency drift/instability from either the bpf or userspace side could result in some oddities in measurement. We may need to run the bpf updates to the tables at a higher frequency to minimize potential error for that case. This is even more reason to keep the bpf lightweight - but need to consider the implications of the old and new approach with regard to this.